golang/cmd/grafana-proxy/main.go

```diff
@@ -15,110 +15,130 @@
-	zap.S().Infof("Ready to proxy connections")
-
-	// Allow graceful shutdown
-	sigs := make(chan os.Signal, 1)
-	signal.Notify(sigs, syscall.SIGTERM)
-
 	go func() {
-		// before you trapped SIGTERM your process would
-		// have exited, so we are now on borrowed time.
-		//
-		// Kubernetes sends SIGTERM 30 seconds before
-		// shutting down the pod.
-
-		sig := <-sigs
-
-		// Log the received signal
-		zap.S().Infof("Received SIGTERM", sig)
-
-		// ... close TCP connections here.
-		ShutdownApplicationGraceful()
-
+		err := rs.factorySrv.ListenAndServe()
+		if !errors.Is(err, http.ErrServerClosed) {
+			zap.S().Fatalf("failed to start factory server: %s", err)
+		}
 	}()
-
-	select {} // block forever
 }
 
-func isShutdownEnabled() healthcheck.Check {
-	return func() error {
-		if shutdownEnabled {
-			return fmt.Errorf("shutdown")
+// Shutdown the factory and healthcheck servers within a 30 second timeout.
+func (rs *restServers) onShutdown() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	var err error
+	// Servers shouldn't be nil, but better safe than sorry.
+	if rs.factorySrv != nil {
+		fErr := rs.factorySrv.Shutdown(ctx)
+		if fErr != nil {
+			err = fmt.Errorf("failed to shutdown factory server: %w", fErr)
+		} else {
+			zap.S().Info("factory http server successfully shut down")
+		}
+	}
+	if rs.healthSrv != nil {
+		hErr := rs.healthSrv.Shutdown(ctx)
+		if hErr != nil {
+			// Append to existing error if there is one.
+			errors.Join(err, fmt.Errorf("failed to shutdown healthcheck server: %w", hErr))
+		} else {
+			zap.S().Info("healthcheck http server successfully shut down")
 		}
-		return nil
 	}
+	return err
 }
 
-func ShutdownApplicationGraceful() {
-	zap.S().Infof("Shutting down application")
-	shutdownEnabled = true
+type Factory struct {
+	APIKey         string // FACTORYINPUT_KEY
+	User           string // FACTORYINPUT_USER
+	BaseURL        string // FACTORYINPUT_BASE_URL
+	InsightBaseURL string // FACTORYINSIGHT_BASE_URL
+}
 
-	zap.S().Infof("Successful shutdown. Exiting.")
-	os.Exit(0)
+// Initialize the factory struct by reading the required environment variables set in factoryEnvVars.
+// Returns an error if any of the environment variables are missing.
+func initFactory() (Factory, error) {
+	var f Factory
+	formatUrl := func(url string) string {
+		if !strings.HasSuffix(url, "/") {
+			url = fmt.Sprintf("%s/", url)
+		}
+		return url
+	}
+
+	factoryEnvVars := []string{"FACTORYINPUT_KEY", "FACTORYINPUT_USER", "FACTORYINPUT_BASE_URL", "FACTORYINSIGHT_BASE_URL"}
+	for _, envVar := range factoryEnvVars {
+		val, err := env.GetAsString(envVar, true, "")
+		if err != nil {
+			return f, err
+		}
+
+		switch envVar {
+		case "FACTORYINPUT_KEY":
+			f.APIKey = val
+		case "FACTORYINPUT_USER":
+			f.User = val
+		case "FACTORYINPUT_BASE_URL":
+			f.BaseURL = formatUrl(val)
+		case "FACTORYINSIGHT_BASE_URL":
+			f.InsightBaseURL = formatUrl(val)
+		}
+	}
```

Comment:
Please add a check if all 4 env variables have been set

Reply:
Is it necessary, though? On line 127, I try to get the value associated with the env variable of the current iteration, and if not available, I throw an error, which is then caught in the main function. So if the loop succeeds, the `Factory` struct should be already properly Initialized.

Reply:
true, have overlooked that